src/differentials/thunks.jl

```diff
@@ -9,7 +9,9 @@ Base.Broadcast.broadcastable(x::AbstractThunk) = broadcastable(unthunk(x))
 end
 
 @inline function Base.iterate(::AbstractThunk, (val, state))
-    element, new_state = iterate(val, state)
+    next = iterate(val, state)
+    next isa Nothing && return nothing
+    element, new_state = next
     return element, (val, new_state)
 end
```

Comment:
we may want to rename `val` to `obj` or `underlying_object` or something.
so it is clear that it is the object actually being iterated rather than being a value from the object being iterated